pmd-core/src/main/java/net/sourceforge/pmd/annotation/DeprecatedUntil700.java

```diff
@@ -0,0 +1,14 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.annotation;
+
+/**
+ * Tags a deprecated member that should not be removed before PMD 7.0.0.
+ * Such members were made deprecated on the PMD 7 development branch and
+ * may be kept for backwards compatibility on the day of the PMD 7 release,
+ * because the replacement API cannot be backported to PMD 6.
+ */
+public @interface DeprecatedUntil700 {
```

Comment:
I was thinking, maybe we could ease the transition to PMD 7 by not removing everything right away. Eg maybe we could have a 7.0.0 "pre-release" with some deprecated API (and already a lot of removed/changed APIs), then the real release without those deprecated members. In my mind, keeping around a deprecated member is only appropriate when it's extremely easy (like not more than have a deprecated method that delegates to the new impl).

So we'd effectively need a distinction between 
* "deprecated since 6.0": can be removed before releasing 7.0 without thinking twice about it
* "deprecated since 7.0": the replacement API doesn't exist in 6.x versions, and it's easy to keep around for compatibility

Those visitors are not the best examples for the value-added of this distinction. Let's look for example, at the traversal methods on the `Node` interface. Since we have node streams, methods like `Node::findChildrenOfType` or `Node::findDescendantsOfType` should be removed I think. Node streams are better in that they're lazy and can be chained better.

But we can't deprecate those methods in a 6.x release, because there's no replacement API available yet. Most of them are very dependent on too. For the 7.0.0 release we have several options:
1. Remove them before 7.0.0
2. Keep them in 7.0.0 and deprecate them for removal in 8.0.0. This will make the transition simpler for users, but will give them a very long lifetime (which to us, translates to technical debt) 
3. Keep them in a 7.0.0 "pre-release" and remove them on 7.0.0. This makes the transition simpler for users, at minimal cost to us.

Keeping them on the 7.0.x branch until the release comes at nearly zero cost, for now they're even necessary, since a lot of our codebase still uses them. So I don't really see why we shouldn't extend this "grace period" to other clients. Removing them in 7.0 without transition would make an upgrade to 7.0 harder than it *needs to be*, given that it's probably already going to be challenging.

Am I making sense? What do you think of the idea of such a "pre-release"?

> Alternatively, we could also document, what needs to be done, in the migration guide. I've seen now the place, where you used the new annotation. So we could use the new annotation as a marker/reminder for us, that this needs to be part of the migration guide.

If we don't do a 7.0.0 "pre-release" like that, then yes, I think a migration guide, and using this annotation to help, would be a good compromise too. The point of the annotation to give us time, we don't have to commit to one outcome or another just yet.

---

`DeprecatedUntil700` hints at doing a pre-release. Regardless of whether we go that way or not, maybe `DeprecatedSince700` would be a better name. It doesn't tell us to keep them until 7.0.0, just reminds us that they're deprecated, but that there's no transitional API on master. So it would remind us to
* if possible, add such a transitional API before wrapping up the 6.x release cycle, and remove them before 7.0.0
* otherwise, mention it in a migration guide/ release notes of 7.0.x

Reply:
Thanks for the explanations, I think, I understand now the background. For now, we can of course leave these annotations in and decide later, what we really do.

I'm not sure yet, whether the pre-release will really help the users... e.g. I could imagine, they update to 7.0.0-pre and they see: "ok, it works, lot's of deprecations, but still compiles", then they try out 7.0.0 and see "hm... compiler errors - I don't have time to look at them, let's stay at 7.0.0-pre, since this still works"... ok, it would be the problem of the users, if they don't do the full migration. Or in other words: compile errors can't be ignored, deprecations can't.

In general, I would keep the deprecations only in the pre-release and try to avoid to extend the lifespan of these deprecations - so that we really delete the old stuff with the final version of 7.0.0.

Reply:
As we discussed, we keep the new annotation `@DeprecatedUntil700` and decide later, whether migration doc is enough or whether we can easily create a compatibility-helping pre-release. There are also other candidates, where this could be useful, e.g. in the `Node` / AST API (Node vs. NodeStream).